### ISSUE TITLE
scaling Mel Spectrogram output for Wavenet Vocoder

### ISSUE THREAD

**Issue by G-Wang** [NVIDIA/tacotron2#24]
Hello,

First of all thanks for the nice Tacotron 2 implementation.

I'm trying to use the trained Tacotron 2 outputs as inputs to r9r9's Wavenet vocoder. However his pre-trained wavenet works on scaled Mel Spectrogram between [0, 1].

What is the range for this tacotron 2 implementation, I'm having a hard time finding this out to use it for scaling.

For reference, this is r9r9's normalization function that he applies to the Mel Spectrogram before using it for training, which scales it between 0 and 1:

`def _normalize(S):
    return np.clip((S - hparams.min_level_db) / -hparams.min_level_db, 0, 1)`

**Comment by rafaelvalle** [comment]
Our [dynamic range compression] (https://github.com/NVIDIA/tacotron2/blob/master/audio_processing.py#L78) just apply the log on clamped values. We also provide a dynamic range decompression function [here](https://github.com/NVIDIA/tacotron2/blob/master/audio_processing.py#L87)

I think the code below is what you're looking for.

```
mel = torch.load(conditional_path)
mel = dynamic_range_decompression(mel)
mel = mel.cpu().numpy()
mel = mel.transpose()

mel = audio._amp_to_db(mel) - hparams.ref_level_db
if not hparams.allow_clipping_in_normalization:
    assert mel.max() <= 0 and mel.min() - hparams.min_level_db >= 0
mel = audio._normalize(mel)
```

**Comment by rafaelvalle** [comment]
Curious to hear your samples and to know if you trained with the most recent code that updates the input of the attention and decoder.

**Comment by G-Wang** [comment]
Thanks, that worked, (my model still needs some more training), I'm starting another run with the updated attention repository, will post back here with the entire pipeline (+ wavenet vocoder) once that's done.

**Comment by yliess86** [comment]
Hi @G-Wang  ! Have you gone further with your training ? I am currently doing the same thing as you and the voice I get from Wavenet sounds like if it has the flu. Did you managed to have good results ?

**Comment by rafaelvalle** [comment]
@yliess86 can you share the audio and the mel-spectrogram that sounds like it has the flu here?

**Comment by G-Wang** [comment]
Hello @yliess86, due to my limited compute, I had to set the batch size pretty low to fit everything into the GPU (batch_size of 18), the network's loss was stuck around 0.63 for quite a while, so I didn't continue. however that was older version of the code. I've started running a new training with the latest code since yesterday.

But note that the solution @rafaelvalle  provided works no problem once your network has been trained, as I took the ground truth mel spectrogram data that Tacatron-2 trains on and it has very good quality on R9R9's wavenet vocoder (using the above code)

**Comment by yliess86** [comment]
@G-Wang Ok thank you.
@rafaelvalle Sure! I will share audio and mel-spec today.

**Comment by yliess86** [comment]
Here are the text, the corresponding audio and the corresponding mel-spectrogram:
- 'This is an example of text to speech synthesis after 9 days training. This may sound awful, but it is a start.'
- [Audio](https://instaud.io/2jfI)
![Mel-Spec](https://image.ibb.co/feuJVJ/mel_spec_0.png)

**Comment by rafaelvalle** [comment]
@yliess86 Can you share the mel-spectrogram file?

**Comment by yliess86** [comment]
@rafaelvalle I plugged the output of the tacotron2 to the convertion pipline you described above and finally plugged it into the r9y9 wavenet. The image you can see is the plt plot I did just before the convertion. Do you want me to give you the output (mel-spec) into a .npy file or other ?

**Comment by rafaelvalle** [comment]
Yes, please do share the mel-spec into a torch or npy file.

**Comment by yliess86** [comment]
Here is the file: [Mel-Spec](https://1drv.ms/u/s!Avn9GxYVpH4aa4396i4SY3aNtMw)

**Comment by rafaelvalle** [comment]
@yliess86 The model that produced this mel-spectrogram was not trained on LJ Speech dataset, right?

**Comment by yliess86** [comment]
It was. I trained the model with the LJ Speech dataset. This mel Spec is the result of 70000 iterations on this dataset.

**Comment by rafaelvalle** [comment]
That's unexpected. Did you train using the default params?

**Comment by yliess86** [comment]
Yes, I just changed the batch size to 24. I'll try to download it again and retrain the model. It was the LJ Speech dataset from the link given on the rayhan repository so maybe this is not exactly the same.

**Comment by rafaelvalle** [comment]
What were the training and validation loss of the model used to produce that mel spectrogram?

**Comment by yliess86** [comment]
The training loss was between 0.3 and 0.5 and the validation one was 0.46

**Comment by rafaelvalle** [comment]
Don't retrain the model yet. We're seeing a similar behavior and are working on it.

**Comment by mrgloom** [comment]
Is it possible to use pretrained models from https://github.com/r9y9/wavenet_vocoder with https://github.com/NVIDIA/tacotron2 ? or it should be retrained?
